File: Unidad 3/Ejercicios/MetodoDeConvolucion.java
import java.util.Random;

public class MetodoDeConvolucion {

    private static double generateExponential(Random random, double lambda) {
        double media = 1.0 / lambda; 
        double U = random.nextDouble();
        
        return media * (-Math.log(1.0 - U));
    }

    public static void main(String[] args) {
        
        Random random = new Random(505); // 
        int N = 10000;                  
        int k = 4;                      
        double lambda = 0.5;          
        
        double[] varErlangConvolucion = new double[N];

        System.out.println("--- Método de Convolución (Distribución Erlang) ---");
        System.out.printf("Generando %d muestras de Erlang(k=%d, λ=%.2f)\n\n", N, k, lambda);
        
        for (int i = 0; i < k; i++) {
          
            for (int j = 0; j < N; j++) {
                double Y_i = generateExponential(random, lambda);
              
                varErlangConvolucion[j] += Y_i;
            }
        }

        double mediaTeorica = (double) k / lambda; 
        

        double sumaTotal = 0;
        for (double x : varErlangConvolucion) {
            sumaTotal += x;
        }
        double mediaGenerada = sumaTotal / N;
        
        System.out.println("Primeros 5 valores generados (X = Y1 + ... + Yk):");
        for (int i = 0; i < 5; i++) {
            System.out.printf("  %.4f\n", varErlangConvolucion[i]);
        }
        
        System.out.println("\n--- Resumen Estadístico ---");
        System.out.printf("Media Teórica (k/λ): %.4f\n", mediaTeorica);
        System.out.printf("Media Muestral Generada: %.4f\n", mediaGenerada);
    }
}

File: Unidad 3/Ejercicios/MetodoDeConvolucion.py
import numpy as np
import matplotlib.pyplot as plt

np.random.seed(999)

N = 10000        
k = 4            
lambda_tasa = 0.5 
scale = 1 / lambda_tasa 

var_erlang_convolucion = np.zeros(N)

print(f"Generando {N} muestras de Erlang({k}, λ={lambda_tasa}) mediante convolución...")

for i in range(k):
    
    Y_i = np.random.exponential(scale=scale, size=N)
    
    var_erlang_convolucion += Y_i

var_erlang_nativa = np.random.gamma(shape=k, scale=scale, size=N)

print("\n--- Resultados de Muestras Generadas ---")
print(f"Primeros 5 valores (Convolución): {var_erlang_convolucion[:5].round(4)}")
print(f"Primeros 5 valores (NumPy Nativo): {var_erlang_nativa[:5].round(4)}")

plt.figure(figsize=(7, 5))
plt.hist(var_erlang_convolucion, bins=50, density=True, alpha=0.7, color='darkcyan', label='Convolución (Suma de Exponenciales)')
plt.hist(var_erlang_nativa, bins=50, density=True, alpha=0.5, color='red', label='NumPy Nativo (Gamma)')
plt.title(f'Distribución Erlang(k={k}, $\\lambda$={lambda_tasa})')
plt.xlabel('Valores')
plt.ylabel('Densidad')
plt.legend()
plt.grid(axis='y', alpha=0.5)
plt.show()